### PR TITLE
Parse -Xmso from OPENJ9_JAVA_OPTIONS for main thread

### PR DESCRIPTION
**Pull request by pshipton**
Issue https://github.com/eclipse-openj9/openj9/issues/16105

Backport of https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/644, the JLI helper functions, and parts of args.c applicable to jdk8.

### CODE REVIEW

**Comment by pshipton** [comment]
I didn't add IBM copyrights to a couple of files where the changes are all copies of upstream code.

**Comment by keithc-ca** [comment]
> changes are all copies of upstream code

Which version did you use to source those changes?

**Comment by pshipton** [comment]
> Which version did you use to source those changes?

The head stream, openj9-openjdk-jdk.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by pshipton** [comment]
Addressed the review comments.

File: jdk/src/share/bin/jli_util.h [hunk 1]
```diff
@@ -101,4 +107,32 @@ void     JLI_TraceLauncher(const char* fmt, ...);
...
+/* e must be JLI_MemFree-able */
+void JNICALL
+JLI_List_add(JLI_List l, char *e);
+
+/* a copy is made out of beg */
+void JLI_List_addSubstring(JLI_List l, const char *beg, size_t len);
+char *JLI_List_combine(JLI_List sl);
+
+jboolean JNICALL
+JLI_ParseOpenJ9ArgsFromEnvVar(JLI_List args, const char *var_name);
```
Comment (keithc-ca) [inline jdk/src/share/bin/jli_util.h:136]: `JNICALL` is only necessary for exported functions; suggest you remove them from the declarations here, and from the function definitions.

File: jdk/src/share/bin/jli_util.c [hunk 2]
```diff
@@ -119,3 +119,77 @@ JLI_StrCCmp(const char *s1, const char* s2)
 {
    return JLI_StrNCmp(s1, s2, JLI_StrLen(s2));
 }
+
+JNIEXPORT JLI_List JNICALL
+JLI_List_new(size_t capacity)
```
Comment (keithc-ca) [inline jdk/src/share/bin/jli_util.c:124]: The definitions must be consistent with `jli_util.h` (remove `JNIEXPORT` and `JNICALL`).

File: jdk/src/share/bin/args.c [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+    return JLI_StrCmp(arg, "-jar") == 0 ||
+           JLI_StrCmp(arg, "-h") == 0 ||
+           JLI_StrCmp(arg, "-?") == 0 ||
+           JLI_StrCmp(arg, "-help") == 0 ||
+           JLI_StrCmp(arg, "-X") == 0 ||
+           JLI_StrCmp(arg, "-version") == 0 ||
+           JLI_StrCmp(arg, "-fullversion") == 0;
+}
+
+jboolean JNICALL
```
Comment (keithc-ca) [inline jdk/src/share/bin/args.c:383]: Please remove `JNICALL` here.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by pshipton** [comment]
Updated. If there is a whitespace issue I'll fix it later.

File: jdk/src/share/bin/jli_util.h [hunk 4]
```diff
@@ -101,4 +107,32 @@ void     JLI_TraceLauncher(const char* fmt, ...);
...
+struct JLI_List_
+{
+    char **elements;
+    size_t size;
+    size_t capacity;
+};
+typedef struct JLI_List_ *JLI_List;
+
+JLI_List
+JLI_List_new(size_t capacity);
```
Comment (keithc-ca) [inline jdk/src/share/bin/jli_util.h:122]: Perhaps this declaration, and the ones beginning on lines 128 and 135, should be on one line like the others in this file.

Comment (pshipton) [reply jdk/src/share/bin/jli_util.h]: Updated.

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**